### PR TITLE
quictls 3.0.9 (new formula)

### PR DESCRIPTION
**Pull request by flier**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----
The latest `NodeJS` (#89068 #89459) and `Haproxy` v2.8 use [quictls](https://github.com/quictls/openssl) to [support the QUIC protocol](https://www.openssl.org/blog/blog/2020/02/17/QUIC-and-OpenSSL/).
The formula is is a copy of openssl, so can't pass the `brew audit` either.

```sh
$ brew audit --strict --new-formula --online quictls
quictls
  * line 90, col 5: Formulae in homebrew/core (except e.g. cryptography, libraries) should not run build-time checks
  * GitHub fork (not canonical repository)
```

Can we temporarily turn off this warning to run through the functional tests for each platform ?

### CODE REVIEW

**Comment by SMillerDev** [comment]
These should be two separate PRs

**Comment by Bo98** [comment]
We're phasing out OpenSSL 1.1.1 ahead of its upcoming EOL, so I don't think we should be adding a 1.1.1 for this.

File: Formula/quictls.rb [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+  sha256 "7b76b0968642a3929fabc96acd28302f631d339a90a91fd434cde425a3ec6838"
+  license "Apache-2.0"
+
+  keg_only :shadowed_by_macos, "macOS provides LibreSSL"
+
+  depends_on "ca-certificates"
+
+  on_linux do
+    keg_only "it conflicts with the `openssl@3` formula"
+
```
Comment (carlocab) [inline Formula/quictls.rb:15]:
```suggestion
  keg_only "it conflicts with OpenSSL"

  depends_on "ca-certificates"

  on_linux do
```
Let's just have on `keg_only` reason here.

File: Formula/quictls.rb [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+      url "https://cpan.metacpan.org/authors/id/B/BI/BINGOS/ExtUtils-MakeMaker-7.70.tar.gz"
+      mirror "http://cpan.metacpan.org/authors/id/B/BI/BINGOS/ExtUtils-MakeMaker-7.70.tar.gz"
+      sha256 "f108bd46420d2f00d242825f865b0f68851084924924f92261d684c49e3e7a74"
+    end
+  end
+
+  link_overwrite "bin/c_rehash", "bin/openssl", "include/openssl/*"
+  link_overwrite "lib/libcrypto*", "lib/libssl*"
+  link_overwrite "lib/pkgconfig/libcrypto.pc", "lib/pkgconfig/libssl.pc", "lib/pkgconfig/openssl.pc"
+  link_overwrite "share/doc/openssl/*", "share/man/man*/*ssl"
```
Comment (carlocab) [inline Formula/quictls.rb:38]:
```suggestion
```
We shouldn't do this since this is keg-only.

File: Formula/quictls.rb [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    end
+    args
+  end
+
+  def install
+    if OS.linux?
+      ENV.prepend_create_path "PERL5LIB", libexec/"lib/perl5"
+
+      %w[ExtUtils::MakeMaker Test::Harness Test::More].each do |r|
+        resource(r).stage do
```
Comment (carlocab) [inline Formula/quictls.rb:66]:
```suggestion
      resources.each do |r|
        r.stage do
```

File: Formula/quictls.rb [hunk 4]
```diff
@@ -0,0 +1,132 @@
+class Quictls < Formula
+  desc "This is a fork of OpenSSL to enable QUIC"
```
Comment (carlocab) [inline Formula/quictls.rb:2]:
```suggestion
  desc "TLS/SSL and crypto library with QUIC APIs"
```
Let's use the description from the repo.

File: Formula/quictls.rb [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+  homepage "https://github.com/quictls/openssl"
+  url "https://github.com/quictls/openssl/archive/refs/tags/openssl-3.0.9-quic1.tar.gz"
+  version "3.0.9"
```
Comment (carlocab) [inline Formula/quictls.rb:5]: Should this be
```suggestion
  version "3.0.9-quic1"
```
?

**Comment by flier** [comment]
@carlocab Tests still fail because of full audit

**Comment by carlocab** [comment]
Yes, I can see that, thanks.

**Review by carlocab — APPROVED** [review]
I'm ok with merging this as-is.

We can add this to the [allowlist](https://github.com/Homebrew/homebrew-core/blob/e4d399d97c69741901935271d4f167c6c1269880/style_exceptions/make_check_allowlist.json) to fix the `make check`/`make test` audit failure, but the fork failure can't be made to go away here.

Note, however, that I'm only ok with merging this as long as we **do not** use this as a dependency for other formulae. All formulae should continue to use upstream OpenSSL as long as that continues to be supported.

**Comment by flier** [comment]
> I'm ok with merging this as-is.
> 
> We can add this to the [allowlist](https://github.com/Homebrew/homebrew-core/blob/e4d399d97c69741901935271d4f167c6c1269880/style_exceptions/make_check_allowlist.json) to fix the `make check`/`make test` audit failure, but the fork failure can't be made to go away here.
> 
> Note, however, that I'm only ok with merging this as long as we **do not** use this as a dependency for other formulae. All formulae should continue to use upstream OpenSSL as long as that continues to be supported.



> We're phasing out OpenSSL 1.1.1 ahead of its upcoming EOL, so I don't think we should be adding a 1.1.1 for this.

@Bo98 We may still need to use OpenSSL 1.1.x for a while until the performance issues are mitigated.

openssl/openssl#17064
openssl/openssl#20171

**Comment by SMillerDev** [comment]
> We may still need to use OpenSSL 1.1.x for a while until the performance issues are mitigated.

OpenSSL 3.x is what it is, performance issues and all and soon it'll be the only supported version and Homebrew plans to abide by that.

**Review by chenrui333 — APPROVED** [review]

**Review by carlocab — APPROVED** [review]
Apologies for the delay here. I was waiting for Homebrew/brew#15614 to land. There's some extra changes needed, but I'll follow-up on those. Let's get this merged.

**Pull request merged**